src/webnn/native/dml/DMLUtils.cpp

```diff
@@ -52,42 +69,395 @@ namespace webnn::native::dml {
+            mDescriptorHeap->GetCPUDescriptorHandleForHeapStart();
+        mBindingTableDesc.GPUDescriptorHandle =
+            mDescriptorHeap->GetGPUDescriptorHandleForHeapStart();
+        // The size of the binding table, in descriptors. This is the maximum number of
+        // descriptors that DirectML is permitted to write, from the start of both the
+        // supplied CPU and GPU descriptor handles.
+        mBindingTableDesc.SizeInDescriptors = descriptorCount;
+        WEBNN_CHECK(mDevice->CreateBindingTable(&mBindingTableDesc, IID_PPV_ARGS(&mBindingTable)));
+
+        // Initialize constant inputs.
+        auto inputNodes = compiledGraph->inputNodes;
+        for (auto& inputNode : inputNodes) {
+            if (inputNode->type == NodeType::ConstantInput) {
+                uint64_t offset =
+                    RoundUpToMultiple(mConstantInputsResourceSize,
+                                      static_cast<uint64_t>(DML_MINIMUM_BUFFER_TENSOR_ALIGNMENT));
+                mConstantInputsResourceSize = offset + inputNode->byteLength;
+            } else {
+                uint64_t offset =
+                    RoundUpToMultiple(mNonConstantInputsResourceSize,
+                                      static_cast<uint64_t>(DML_MINIMUM_BUFFER_TENSOR_ALIGNMENT));
+                mNonConstantInputsResourceSize = offset + inputNode->byteLength;
+            }
+        }
+        // Set the descriptor heap(s).
+        ID3D12DescriptorHeap* descriptorHeaps[] = {mDescriptorHeap.Get()};
+        mCommandList->SetDescriptorHeaps(ARRAYSIZE(descriptorHeaps), descriptorHeaps);
+
+        CreateResourcesForCompiledOperatorInitializer();
+        BindTemporaryResource();
+        BindPersistentResource();
+
+        if (mConstantInputsResourceSize) {
+            std::vector<DML_BUFFER_BINDING> inputBufferBinding(inputNodes.size());
+            FillUploadResourceAndInputBindings(inputBufferBinding, inputNodes);
+            // Copy buffer from uploadResource to inputResource.
+            CopyBufferRegion(mUploadResource, mInputResource, mConstantInputsResourceSize,
+                             D3D12_RESOURCE_STATE_COPY_DEST);
+
+            DML_BUFFER_ARRAY_BINDING inputBufferArrayBinding = {};
+            inputBufferArrayBinding.BindingCount = inputBufferBinding.size();
+            inputBufferArrayBinding.Bindings = inputBufferBinding.data();
+            DML_BINDING_DESC inputBindingDesc{DML_BINDING_TYPE_BUFFER_ARRAY,
+                                              &inputBufferArrayBinding};
+            mBindingTable->BindInputs(1, &inputBindingDesc);
+        }
+
+        // Record execution of the operator initializer.
+        // The command recorder is a stateless object that records Dispatches into an existing
+        // Direct3D 12 command list.
+        WEBNN_CHECK(mDevice->CreateCommandRecorder(IID_PPV_ARGS(&mCommandRecorder)));
+        mCommandRecorder->RecordDispatch(mCommandList.Get(),
```

Comment:
I intend to move the initialization as more as possible into compilation stage to reduce the time cost in execution

Reply:
Multiple calls to ExecuteCommandList actually increases the time cost of execution because the CPU needs more time to prepare the GPU for execution.

Remember, a command-list is like a background thread for the GPU. Except that "thread" is massively parallel.  You need to feed the GPU with MORE work at once to keep that thread busy, not create more "threads" with less work.

Reply:
In this way, for the first time execution, move all of things from initializeGraph to ExecuteGraph except CreateOperatorInitializer and then only call ExecuteCommandList  once.
Although it can indeed feed the GPU with more work at once to reduce time cost, but I am afraid that the additional work moved into ExecuteGraph will cost more time. So the total time cost will be unclear? Maybe we need try and compare.

Reply:
@mingmingtasd 

Batching commands into a single command list should not increase overall work. It decreases CPU work, by not preparing a command-list, and reduces GPU time, since commands within a list do not need to be serialized. It is counter-productive to break them up into multiple command-lists when work is dependent.

Reply:
I tried to understand what you mean, please note that:
1. To align with webnn-spec, we must read back the output data from GPU to CPU, I mean we should get the output data when the ExecuteGraph() is finished, so we must call ExecuteCommandList once in each ExecuteGraph().
2. You mean move most of the codes in InitializeGraph() to ExecuteGraph(), I am afraid the time cost of calling ExecuteGraph() for the first time may be long because we observed that calling InitializeGraph() for the first time is always very slow.

Reply:
Initialize should only be called only once, unless the input (graph) changes. If `ExecuteGraph()` gets slower, after moving `RecordDispatch` there, then you're likely initializing again. The problem is the code lacks a means to track which state the operator (initialized or not) needs to be in for recording so it splits it up over two `ExecuteCommandLists` calls. Note: this issue is completely unrelated to always reading back from the GPU, I will address that offline.